### ISSUE TITLE
Airpods Pro 2nd generation

### ISSUE THREAD

**Issue by Karthinabi** [steam3d/MagicPods-Windows#246]
Add support latest Airpods Pro (2nd generation) 2022

**Comment by jonathanfan-ee** [comment]
+1

**Comment by steam3d** [comment]
Hi. I can't say anything. It difficult to get the AirPods Pro 2 for me. I will follow the news about AirPods Pro 2

**Comment by YongHuM1ng** [comment]
> Hi. I can't say anything. It difficult to get the AirPods Pro 2 for me. I will follow the news about AirPods Pro 2

I have an Airpods Pro 2, is there anything I can do to help?

**Comment by joshhoughton** [comment]
+1, same here, please let me know if there is anything I can do! @steam3d 

Just want to say, amazing app, have been using for the last 1.5 years with my Airpods Pro 1 every single day.

Regretting upgrading to the 2nd gen for this reason! <3

**Comment by MiLeX84** [comment]
Same, currently sad that the app doesn't support AirPods Pro 2nd gen. Please add ASAP :)

**Comment by Koterak** [comment]
Also I have Airpods Pro 2, so if there is anything I can help with, let me know

**Comment by steam3d** [comment]
Guys in my country don't officially sell apple appliances.  When I could get headphones - I will try to add.
If you already have the AirPods Pro 2:
1. Could you screen record the new AirPods Pro 2 animation?
2. Could you run the [DeviceFinder](https://github.com/steam3d/MagicPods-Windows/files/4911413/DeviceFinder_1.zip) before connecting AirPods Pro 2:

1. Run the DeviceFinder
2. Connect Beats Studio Buds
3. Try listen to music
4. Put both headphone to case and close it
5. Send the log file here

**Comment by joshhoughton** [comment]
Hey @steam3d, here you go!

[Animation](https://user-images.githubusercontent.com/6955968/192115745-35290d89-6838-4fb8-91be-853a893c77e6.MP4)

[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9639464/log.txt)


Best of luck!

Josh

**Comment by joshhoughton** [comment]
Also, [the full animation when already paired](https://user-images.githubusercontent.com/6955968/192115840-9712f3a7-29dc-492a-8eef-ca735b7f4230.MP4)

**Comment by XUYinong** [comment]
> Guys in my country don't officially sell apple appliances. When I could get headphones - I will try to add. If you already have the AirPods Pro 2:
> 
> * Could you screen record the new AirPods Pro 2 animation?
> * Could you run the [DeviceFinder](https://github.com/steam3d/MagicPods-Windows/files/4911413/DeviceFinder_1.zip) before connecting AirPods Pro 2:
>   
>   1. Run the DeviceFinder
>   2. Connect AirPods Pro 2
>   3. Try listen to music
>   4. Put both headphone to case and close it
>   5. Send the log file here

if you need more log🙏
[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9640114/log.txt)

**Comment by YongHuM1ng** [comment]
https://user-images.githubusercontent.com/44365990/192132037-fa260101-fc75-483c-afb1-7897cb5c0698.mp4
[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9640189/log.txt)

**Comment by steam3d** [comment]
Looks like AirPods Pro works different because it has a new case. They seem to work, but I can't guarantee that the case won't get in the way. I can try to test the program on your data. Could someone do this?

1. Run the [DeviceFinder](https://github.com/steam3d/MagicPods-Windows/files/4911413/DeviceFinder_1.zip) before connecting AirPods Pro 2:
2. Bring case AirPods Pro with AirPods inside as close as possible to Bluetooth dongle  if you have PC or laptop.
3. Wait 5 sec
4. Open the case with AirPods inside
5. Wait 5 sec
6. Take left AirPod out from the case
7. Wait 5 sec
8. Take right AirPod out from the case
9. Wait 5 sec
10. Put left AirPod in ear 
11. Wait 5 sec
12. Put right AirPod in ear
13. Wait 5 sec
14. Put left AirPod out from ear
15. Wait 5 sec
16. Put right AirPod out from ear
17. Wait 5 sec
18. Put left AirPod into the case
19. Wait 5 sec
20. Put right AirPod into the case
21. Wait 5 sec
22. Close the case
23. Wait 5 sec
24. Upload the log.txt

**Comment by joshhoughton** [comment]
Hey @steam3d, please find the log file attached :)

[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9644264/log.txt)

**Comment by zdravkoggeorgiev** [comment]
Привет, @steam3d  сделал что ты просил в log.txt
[log.zip](https://github.com/steam3d/MagicPods-Windows/files/9644660/log.zip)

1. Trial можно сбросить, когда новая версия въйдет для Airpods Pro 2? Я установил, но не могу попробовать.
2. Кроме визуальный настроек есть и другие подобрения? У меня проблем что ничего не слышно с микрофона Airpods Pro 2. Надеялся, эта прога изправит.

**Comment by steam3d** [comment]
@zdravkoggeorgiev Did you put both AirPods in ears?  Because i did not find it in log files

**Comment by joshhoughton** [comment]
@steam3d just curious, what do you use to decode the hex that comes from the script?

Could you find those events in my log file?

**Comment by steam3d** [comment]
@joshhoughton I did not decode this data. Bluetooth advertisement is byte data. The Apple sends this data across all their devices. So no one know how it works. Because of it the AirPods 1 may works perfectly but AirPods Pro 2 not. 

I checked your log and find that only Left AirPod was in ear.

I have 2 ways now.  I can enable limited support for AirPods Pro 2. EarDetection and popup animation features will not work. I will try to figure out what's wrong when I get AirPods Pro 2.

So you could try to put AirPods in case and put them out and put them in ears multiple times to get more inform log file. I think the new case has Bluetooth module as well, thus AirPods Pro 2 has 3 Bluetooth modules.

**Comment by joshhoughton** [comment]
Hey @steam3d  thanks for this info!

I've just retried:

1. opening case, 
2. putting left AirPod in ear, waiting ~5 seconds, taking out of ear, waiting ~5 seconds [x3]
[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9645582/log.txt)

3. putting left AirPod back in case
4. putting right AirPod in ear, waiting ~5 seconds, taking out of ear, waiting ~5 seconds [x3]
5. putting right AirPod back in case
6. close case

Let me know if you need any more information, happy to help however I can.

**Comment by zdravkoggeorgiev** [comment]
@steam3d 
There could be some issue on my side because I have mention multiple times that puting into ear doesn't work or pushing buttons doesn't work. I can put both earplugs but only one plays music. It happens like 5% of time.

 I will wait a week to see - will there an update come out? If there is no update then probably I have issues with my earplugs.
 
As additional note - there is one additional device which is found in Windows as "stb bsa bluetooth device". I am not sure is this the case. "BTHLE\DEV_402F8641D36E\7&26C4689&0&402F8641D36E"

**Comment by steam3d** [comment]
@joshhoughton I can confirm that Left and Right AirPod was detected. I can also confirm that some data does not match with other my AirPods. Thanks for help.
I will wait some other people who give their data and I will compare with yours

**Comment by XUYinong** [comment]
> @joshhoughton I can confirm that Left and Right AirPod was detected. I can also confirm that some data does not match with other my AirPods. Thanks for help. I will wait some other people who give their data and I will compare with yours

Here is my log file, hope it can be helpful
[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9646331/log.txt)

**Comment by tmalust** [comment]
[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9646399/log.txt)

Mine as well :)

**Comment by steam3d** [comment]
Can someone collect data with closed case?
Connect AirPods Pro to PC and play some music. After it run DeviceFinder on 10-15 sec.

**Comment by YongHuM1ng** [comment]
> 看起来AirPods Pro 2的工作方式不同，因为它有一个新的外壳。他们似乎有效，但我不能保证这个案子不会妨碍。我可以尝试在您的数据上测试程序。有人可以这样做吗？
> 
> 1. 在连接空气豆荚 Pro 2 之前运行[设备查找器](https://github.com/steam3d/MagicPods-Windows/files/4911413/DeviceFinder_1.zip)：
> 2. 如果您有PC或笔记本电脑，请将带有空气豆荚的空气豆荚专业版外壳尽可能靠近蓝牙加密狗。
> 3. 等待 5 秒
> 4. 打开装有空气豆荚的箱子
> 5. 等待 5 秒
> 6. 从箱子里拿出左边的空气舱
> 7. 等待 5 秒
> 8. 从箱子里拿出右一个空气豆荚
> 9. 等待 5 秒
> 10. 将左空气荚放在耳朵里
> 11. 等待 5 秒
> 12. 将正确的空气荚放在耳朵里
> 13. 等待 5 秒
> 14. 将左空气荚从耳朵中取出
> 15. 等待 5 秒
> 16. 将右空气荚从耳朵中取出
> 17. 等待 5 秒
> 18. 将左空气荚放入箱子中
> 19. 等待 5 秒
> 20. 将正确的空气荚放入机箱中
> 21. 等待 5 秒
> 22. 关闭案例
> 23. 等待 5 秒
> 24. 上传日志.txt

[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9647967/log.txt)

**Comment by XUYinong** [comment]
This log file was generated under the condition that:
1. My PC connected to AirPods and the case was closed and put aside.
2. I started playing music.
3. As music was played, I ran DeviceFinder for around 20-30s.
4. Finally, closed the program.
[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9647984/log.txt)

**Comment by steam3d** [comment]
Right. I will try to add support AirPods Pro 2 at end of the week. I enable all features that other AirPods are supported.
I will write here when I upload the update.  If there are any problems, then I will have to wait until I can get the headphones. But I think it will work.
Thanks everyone for your help.

**Comment by steam3d** [comment]
Update is ready. Go to Microsoft Store and manually check the update. Leave feedback how AirPods Pro 2 works. if there is no update please wait couple hours while Microsoft updates information.

**Comment by joshhoughton** [comment]
> I will try to add support AirPods Pro 2 at end of the week.

He lies! Thanks so much Alex @steam3d, seems to be working perfectly, amazing work mate! <3

**Comment by MrTox** [comment]
Thanks for your work! I have tried the newly released APP. All the functions perform good, and the volume adjust gesture works well!

**Comment by TristanBlus** [comment]
The connect animation did not show after the first connection...

**Comment by v0s** [comment]
Airpods Pro 2 appeared in MagicPods after re-installing to 2.0.5.0, but in-ear status doesn't seem to work reliably. 
Paused once when i took out my left airpod; didn't pause when taking out right one; paused when i put the right one into the case; paused when i put both of them into the case and closed it; didn't un-pause a single time. 

Can my portion of DeviceFinder logs help?

**Comment by steam3d** [comment]
I did not understand what do you mind.  Do you try to enable fast search? Do you try to wait after you put out right AirPod?

**Comment by v0s** [comment]
Tried to enable fast search now, no changes. Waited for 10+ seconds after taking out. Still only paused one time during my tests, and didn't pause again when I did the same thing

**Comment by steam3d** [comment]
@v0s 
Ear detection works:
- Two AirPods in ear: Play or Enable Ear Detection
- One Airpod out of ear: Pause
- Two AirPods out of ear: Disable Ear Detection or Pause if previous step skipped.

So anyway i can't check because I still do not have the AirPods Pro 2

**Comment by v0s** [comment]
Yup, my first gen. Airpods Pro work exactly like that. Is there a log maybe where the in-ear status detected by MagicPods gets reflected? Might come handy looking at how MagicPods sees the in-earness of the pods

**Comment by steam3d** [comment]
You can try this:
1. Go to General Settings
2. Slide the Search Radius to 90-120
3. Go to AirPods Pro 2 settings and enable Fast Search
4. Try to use ear detection

Then describe Does EarDetection work like i describe above? Upload the log files from Support page here.

Note. The MagicPods can find other AirPods not yours.

**Comment by Myndmelt** [comment]
Hi, I got my Airpods Pro 2 today, and went to update and test Magic Pods. I see the support currently is experiemental for understandable reasons.

I can connect my airpods to bluetooth on my pc, and connect them in the magic pods app but IF i do get any audio its for about half second and it cuts out. If I close the magic pods app Im still not getting audio.
I then went and updated my BT driver (built into motherboard Intel Wireless Bluetooth(R) adapter.  In the device manager i see my airpods Pro - Find My and Airpods PRo Find My Avrcp Transport.

Did anyone else have to do any tricks to get these to work with regular bt Audio?  The default output is set to 
Airpod's Pro - Find My (stereo) I tried music app and my browser and im not getting anything going. If anyone has any suggestions please let me know as I wish to help test.

**Comment by steam3d** [comment]
@Myndmelt I think it is not the MagicPods issue. Does the AirPods Pro 2 works good with other devices? Any headphones must work good because for them are enough Bluetooth 2.0 for Audio Playback.

**Comment by Myndmelt** [comment]
> @Myndmelt I think it is not the MagicPods issue. Does the AirPods Pro 2 works good with other devices? Any headphones must work good because for them are enough Bluetooth 2.0 for Audio Playback.

I agree I dont think its a magicpods issue either.  Ive tested them only on my iPhone 12 and iPad. I'll see if I can test it on my windows 10 laptop.  I just read a post earlier on airpods subreddit about another user having the win10 issue that I am facing as well. Very weird.

**Comment by steam3d** [comment]
Is ear detection working for anyone? 
I received AirPods Pro 2 and ear detection does not work because Apple closed the way to use AirPods Pro 2 on none apple devices or maybe it is problem of my AirPods Pro 2

**Comment by v0s** [comment]
It works for me very sporadically and single times, couldn't figure out when it fires, but it's very very rare

**Comment by steam3d** [comment]
@v0s Bad news. I am trying to figure out and looks like the AirPods Pro 2 detects that user use none Apple deivce and stop sharing information.

**Comment by joshhoughton** [comment]
Bastards…On 20 Oct 2022, at 14:40, Alexander Maslov ***@***.***> wrote:﻿
@v0s Bad news. I am trying to figure out and looks like the AirPods Pro 2 detects that user use none Apple deivce and stop sharing information.

**Comment by v0s** [comment]
@steam3d what tools do you use to monitor the information from the airpods? Can i use some BLE explorer to try to figure out what they’re sending?

**Comment by steam3d** [comment]
@v0s Yes, you can. But it does not give usable information as I said above, AirPods Pro 2 stopped sharing data except battery info. so if apple decides to update all their headphones, they can do the same.

**Comment by abmark88** [comment]
Any update on this? Here is my log
[log.txt](https://github.com/steam3d/MagicPods-Windows/files/9872400/log.txt)

**Comment by steam3d** [comment]
@akimbo-md Does it works without MagicPods? I do no have such behavior.

**Comment by akimbo-md** [comment]
> 

It doesn't. Same behaviour using the Windows Bluetooth controls. I haven't been able to use my Airpods Pro 2 at all on this machine.

**Comment by steam3d** [comment]
@akimbo-md I heard that some Android users has trouble with connections. Connection drop down after 90 seconds. Maybe it is AirPods Pro 2 issue. I have Intel Bluetooth 4.2 and AIrPods Pro 2 connects and works good.

I as developer does not have access to Bluetooth Core so I can't check what's wrong. But I spent a lot of time to pair the AirPods Pro 2 with my PC. I have no idea what's was wrong but after couple of restart my laptop everything started work.

**Comment by abventorg** [comment]
@steam3d Hi! Thanks a lot for your work on this great app! Is it possible to add support for sending a pause/play command when the AirPods Pro 2 stem is pressed?

**Comment by v0s** [comment]
@abventorg this works for me natively in Windows even without MagicPods. Volume up/down by caressing the stem works as well

**Comment by abventorg** [comment]
@v0s thanks. I tried another media player and saw the pressing stops/resumes the playback!

**Comment by steam3d** [comment]
I tested the new 5B58 firmware. Nothing changed in logic. AirPods Pro 2 still update their status only when connected to Apple devices.

**Comment by GlxryX** [comment]
Are AAP2 usable with MagicPods? Debating on whether or not to get the original AAP or the AAP2 but don't want to get the newer one if functionality with the newer ones is lacking when it comes to the MagicPods application.

**Comment by v0s** [comment]
AAP2 will not work with in-ear detection (i.e. won't pause the music when you take an airpod out), but it will display the battery charge

**Comment by GlxryX** [comment]
@v0s is that a permanent issue?

**Comment by v0s** [comment]
Seems so, @steam3d says AAP2 just don't report that info over bluetooth if they detect they're connected to a non-apple device (somehow)

**Comment by steam3d** [comment]
@GlxryX I recommend not make decision based on MagicPods, because Apple can changes any thing at any time. Of course I keep looking for solutions, but it's very difficult and slow.

I think that AirPods Pro 2 has a bug in their firmware, because I could not figure out why it stops share information about itself when it connected to none-Apple device.

I think there is alternative way to get raw data from AirPods, but Windows Bluetooth stack does not allow the use of low-level Bluetooth protocols.  If it was possible I could add a remap keys.

**Comment by steam3d** [comment]
Does Ear detection work well for someone? 

if not I will disable Ear Detection feature from AirPods Pro 2 next MagicPods update until some thing change and etc, because some users confuse with it. (According to my tests, ear detection does not work).

**Comment by Rodske** [comment]
I can connect with my Airpods Pro 2 - and see per pod battery power. Is there a way to see the power of the case?

**Comment by steam3d** [comment]
@Rodske put at least one AirPod in open case and wait 0-10 seconds.

**Comment by Void48** [comment]
Does Low Latency work with the new AirPods Pro 2?

**Comment by steam3d** [comment]
@Void48 It works, but the name "Low audio latency" maybe be confusing. The MagicPods tries to preverent headphones go to IDLE mode. Because for some Bluetooth modules need from 0.2 - 1.5 second to go out from IDLE mode. For app like Adobe Premier Pro or Adobe Audition it very user useful.

**Comment by Oculus89** [comment]
Just wondering if anything has changed as far as the in ear detection working or not on the AirPods pro 2nd generation

**Comment by steam3d** [comment]
@Oculus89 No. Once you turn on music the AirPods Pro 2 stops sending data when you listen to music on Windows or Android, but if you listen to the music on IOS or MAC, the AirPods send data as usual.

**Comment by steam3d** [comment]
Added native support of AirPods Pro 2. 
> [MagicAAP driver required](https://help.magicpods.app/fun-magicaap/) and MagicPods 2.0.72.0 and above

## Implemented 
- [x] Accurate battery
- [ ] Rename 
- [x] Noise control
- [ ] Press and hold siri
- [x] Press and hold anc
- [x] Mute & unmute
- [x] End call
- [x] Automatic ear detection (Insert and remove from ear)
- [ ] Automatic ear detection (On/OFF)
- [ ] Spatial audio
- [x] Microphone
- [ ] Enable charging case sounds
- [ ] Optimised battery charging
- [ ] Find my network
- [x] Press speed
- [x] Press and hold duration
- [x] Noise cancellation with One AirPod
- [ ] Spatial audio head tracking
- [x] Tune volume
- [ ] Volume swipe
- [ ] Volume swipe length
- [ ] Firmware info

**Comment by v0s** [comment]
Wow, congrats on the great reverse engineering work
